File: civic_api/app/services/superadmin_service.py
from sqlalchemy.orm import Session

from app.auth import hash_password
from app.models import Complaint, Department, User


def get_all_users(db: Session) -> list[User]:
    return db.query(User).all()


def promote_user(db: Session, user_id: int) -> User:
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise ValueError("User not found")
    if user.role == "SUPER_ADMIN":
        raise ValueError("Cannot modify SUPER_ADMIN")
    if user.role != "DEPARTMENT_ADMIN":
        user.role = "DEPARTMENT_ADMIN"
        db.commit()
        db.refresh(user)
    return user


def demote_user(db: Session, user_id: int) -> User:
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise ValueError("User not found")
    if user.role == "SUPER_ADMIN":
        raise ValueError("Cannot demote SUPER_ADMIN")
    if user.role != "USER":
        user.role = "USER"
        db.commit()
        db.refresh(user)
    return user


def delete_user(db: Session, user_id: int):
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise ValueError("User not found")
    if user.role == "SUPER_ADMIN":
        raise ValueError("Cannot delete SUPER_ADMIN")
    db.delete(user)
    db.commit()


def get_all_departments(db: Session) -> list[Department]:
    return db.query(Department).all()


def create_department(db: Session, name: str, description: str | None) -> Department:
    department = Department(name=name, description=description)
    db.add(department)
    db.commit()
    db.refresh(department)
    return department


def delete_department(db: Session, department_id: int):
    department = db.query(Department).filter(Department.id == department_id).first()
    if not department:
        raise ValueError("Department not found")

    assigned = db.query(User).filter(User.department_id == department_id).first()
    if assigned:
        raise ValueError("Department has assigned admins")

    complaints = db.query(Complaint).filter(Complaint.department_id == department_id).first()
    if complaints:
        raise ValueError("Department has complaints")

    db.delete(department)
    db.commit()


def create_department_admin(
    db: Session, name: str, email: str, password: str, department_id: int
) -> User:
    if db.query(User).filter(User.email == email).first():
        raise ValueError("Email already registered")

    department = db.query(Department).filter(Department.id == department_id).first()
    if not department:
        raise ValueError("Department not found")

    user = User(
        name=name,
        email=email,
        password=hash_password(password),
        role="DEPARTMENT_ADMIN",
        department_id=department_id,
    )
    db.add(user)
    db.commit()
    db.refresh(user)
    return user


def get_all_complaints(db: Session) -> list[Complaint]:
    return db.query(Complaint).all()


def delete_complaint(db: Session, complaint_id: int):
    complaint = db.query(Complaint).filter(Complaint.id == complaint_id).first()
    if not complaint:
        raise ValueError("Complaint not found")
    db.delete(complaint)
    db.commit()

File: civicbackend/src/main/java/com/civic/civicbackend/service/SuperAdminService.java
package com.civic.civicbackend.service;

import com.civic.civicbackend.model.Complaint;
import com.civic.civicbackend.model.Department;
import com.civic.civicbackend.model.User;
import com.civic.civicbackend.repository.ComplaintRepository;
import com.civic.civicbackend.repository.DepartmentRepository;
import com.civic.civicbackend.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import com.civic.civicbackend.dto.AdminRequest;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class SuperAdminService {

    private final UserRepository userRepository;
    private final DepartmentRepository departmentRepository;
    private final ComplaintRepository complaintRepository;
private final PasswordEncoder encoder;
    
    public SuperAdminService(UserRepository userRepository, 
                        DepartmentRepository departmentRepository,
                        ComplaintRepository complaintRepository,
                        PasswordEncoder encoder) {
    this.userRepository = userRepository;
    this.departmentRepository = departmentRepository;
    this.complaintRepository = complaintRepository;
    this.encoder = encoder;
}

    // GET /users - Return all users
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    // PUT /promote/{userId} - Change role to DEPARTMENT_ADMIN
    public User promoteUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        // Cannot modify SUPER_ADMIN
        if ("SUPER_ADMIN".equals(user.getRole())) {
            throw new RuntimeException("Cannot modify SUPER_ADMIN");
        }
        
        // If already DEPARTMENT_ADMIN, do nothing
        if ("DEPARTMENT_ADMIN".equals(user.getRole())) {
            return user;
        }
        
        // Set role to DEPARTMENT_ADMIN
        user.setRole("DEPARTMENT_ADMIN");
        return userRepository.save(user);
    }

    // PUT /demote/{userId} - Change role to USER
    public User demoteUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        // Cannot demote a SUPER_ADMIN
        if ("SUPER_ADMIN".equals(user.getRole())) {
            throw new RuntimeException("Cannot demote SUPER_ADMIN");
        }
        
        // Only allow demoting DEPARTMENT_ADMIN to USER
        if ("USER".equals(user.getRole())) {
            return user;
        }
        
        user.setRole("USER");
        return userRepository.save(user);
    }

    // GET /departments - Return all departments
    public List<Department> getAllDepartments() {
        return departmentRepository.findAll();
    }

    // POST /departments - Create new department
    public Department createDepartment(String name, String description) {
        Department department = new Department();
        department.setName(name);
        department.setDescription(description);
        return departmentRepository.save(department);
    }

    // DELETE /departments/{id} - Delete department
    public void deleteDepartment(Long id) {
        Department department = departmentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Department not found"));
        
        // Check if any user has this department assigned
        List<User> allUsers = userRepository.findAll();
        for (User user : allUsers) {
            if (user.getDepartment() != null && user.getDepartment().getId().equals(id)) {
                throw new RuntimeException("Department has assigned admins");
            }
        }
        
        // Check if any complaint belongs to this department
        List<Complaint> complaints = complaintRepository.findByDepartment(department);
        if (!complaints.isEmpty()) {
            throw new RuntimeException("Department has complaints");
        }
        
        departmentRepository.deleteById(id);
    }

    // GET /complaints - Return all complaints
    public List<Complaint> getAllComplaints() {
        return complaintRepository.findAll();
    }

    // DELETE /complaints/{id} - Delete any complaint
    public void deleteComplaint(Long id) {
        if (!complaintRepository.existsById(id)) {
            throw new RuntimeException("Complaint not found");
        }
        complaintRepository.deleteById(id);
    }

    public User createDepartmentAdmin(AdminRequest request) {

    // 🔍 Check email already exists
    if (userRepository.findByEmail(request.getEmail()) != null) {
        throw new RuntimeException("Email already registered");
    }

    // 📌 Get department
    Department department = departmentRepository.findById(request.getDepartmentId())
            .orElseThrow(() -> new RuntimeException("Department not found"));

    // 👤 Create new user
    User user = new User();
    user.setName(request.getName());
    user.setEmail(request.getEmail());

    // 🔐 Encrypt password (VERY IMPORTANT)
    user.setPassword(encoder.encode(request.getPassword()));

    // 🎯 Assign role
    user.setRole("DEPARTMENT_ADMIN");

    // 🔗 Link department
    user.setDepartment(department);

    return userRepository.save(user);
}
}

